### ISSUE TITLE
Add "replace tear-off with function literal" refactor

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#55261]
Consider the following:

```dart
void fn() {}

Button(
  onTap: fn,
)
```

It'd be great to have a refactor that allows expanding the tear-off into:

```dart
void fn() {}

Button(
  onTap: () => fn(),
)
```

This would be the opposite to "replace function literal with tear-off".

This would enable users to more easily "eject" a tear-off to add custom logic.

**Comment by bwilkerson** [comment]
Absolutely. It's now assigned to you.

**Comment by TheComputerM** [comment]
This does not seem to associate with any lint rules, so I am thinking to making this a refactor assist rather than a quick fix.

**Comment by bwilkerson** [comment]
Yes, I believe that an assist is the right approach here. Assists are also written using a `CorrectionProducer`, so the difference between a fix and an assist is fairly small.

**Comment by TheComputerM** [comment]
Also how do I solve the problem of arguments when expanding, for example take this:
```dart
bool isPair(int a, [int b = 0]) => (a + b) % 2 == 0;
final finalList = <int>[1,2,3,4,5,6,7];
finalList.where(isPair);
```

when `isPair` is expanded, what should it's parameters be? do we look at what the parent function takes as input or do we take a look at what *isPair* expects as parameters, what about optional arguments?

**Comment by bwilkerson** [comment]
> Also how do I solve the problem of arguments when expanding

If I understand the original request correctly, I believe that what's desired is for the last line of your example to be replace with
```dart
finalList.where((int a, [int b = 0]) => isPair(a, b));
```

In other words, I believe that we want to copy the parameter list of the tear-off and make it be the parameter list of the closure, then pass the parameters of the closure as arguments to the invocation of the original tear-off.

**Comment by TheComputerM** [comment]
Wouldn't that be a nightmare for some cases (like flutter) where class constructures have more than 10-15 parameters, should it only focus on the required ones or ...

**Comment by rrousselGit** [comment]
> If I understand the original request correctly, I believe that what's desired is for the last line of your example to be replace with
> 
> ```dart
> finalList.where((int a, [int b = 0]) => isPair(a, b));
> ```

Yes. And without specifying types unless the "explicit types" lint is enabled

**Comment by bwilkerson** [comment]
> Wouldn't that be a nightmare for some cases (like flutter) where class constructures have more than 10-15 parameters, should it only focus on the required ones or ...

If a tear-off is being used, then the location it's being passed to is typically expecting a function, and the torn-off function has to be a subtype of the required function type. The closure also has to be a subtype of the required type. I don't think we have a lot of choice here. (Modulo one small issue having to do with what 'subtype' means for function types that I think we're better off ignoring.)

**Comment by TheComputerM** [comment]
So just to clarify, the way we should go about it is: when the tear off is being passed as a parameter, we observe the required function type:

```dart
void printGreeting(String msg, {String name = 'John'}) {
    print('$msg, $name!');
}

list.forEach(printGreeting);
// will become
list.forEach((msg) => printGreeting(msg)); // we expand based on what forEach expects
// or this depending on the lint rules
list.forEach((String msg) => printGreeting(msg));
```
This will be the same if we pass the tear off to something like a class constructor.

Otherwise if it just being assigned to another variable then we observe the tearoff's on type:
```dart
void printGreeting(String msg, {String name = 'John'}) {
    print('$msg, $name!');
}
var a = printGreeting;

// will become

var a = (String msg, {String name = 'John'}) => printGreeting(msg, name: name);
```

(PS: sorry was very busy for the past few days so this slipped my mind).

**Comment by TheComputerM** [comment]
My college plans to have a lot of surprise evals in this month of April so I think I will unassign myself from this issue so that anyone else can pick this if they feel free. (I'll take a jab at it again if I am free).

**Comment by lrhn** [comment]
I think it should be enough to add parameters that match the context type, including whether they're optional or not, and omit optional parameters that the context won't use.
At least if there is a context function type, otherwise all parameters should be forwarded. 
(And sometimes that's not possible, if an optional parameter's default value is not expressible. Most likely there will be a context type with no optional parameters.)

**Comment by FMorschel** [comment]
Hey! About the lint, there is [this proposal](https://github.com/dart-lang/linter/issues/5058) I made which could cover some cases.
